compiler/rustc_hir_typeck/src/_match.rs

```diff
@@ -198,7 +197,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
 
         // Next, make sure that we have no type expectation.
         let Some(ret) = hir
-            .find_by_def_id(self.body_id.owner.def_id)
+            .find_by_def_id(self.body_id)
```

Comment:
There is something fishy here: we cannot have both `get_by_def_id(self.body_id)` a `Node::Expr(ExprKind::Block)` and `find_by_def_id(self.body_id) = Some(get_by_def_id(self.body_id))` have a `fn_decl`.

Reply:
Mh, I can not catch what is the possible solution here?

Reply:
You'll probably want to do 
```suggestion
            .find_by_def_id(self.tcx.parent(self.body_id))
```

As I think the `DefId` parent should be the same item we were looking for before?

Reply:
ok, so after deep thinking on this PR and after talking with some people, I think this is what we need!

```diff
+++ b/compiler/rustc_hir_typeck/src/_match.rs
@@ -197,7 +197,7 @@ fn suggest_removing_semicolon_for_coerce(
 
         // Next, make sure that we have no type expectation.
         let Some(ret) = hir
-            .find_by_def_id(self.body_id)
+            .find_by_def_id(self.tcx.hir().body_owned_by(self.body_id).def_id)
             .and_then(|owner| owner.fn_decl())
             .map(|decl| decl.output.span()) else { return; };
         let Expectation::IsLast(stmt) = expectation else {
```

Reply:
this is still unaddressed

Reply:
Sorry forget to apply!

Reply:
Ah no, I did not forget, but I reverted it.

This is the cause of the following UI failure `tests/ui/suggestions/match-with-different-arm-types-as-stmt-instead-of-expr.rs`

```

------------------------------------------
diff of stderr:

22	   |
23	   = note: expected struct `Box<Baz>`
24	              found struct `Box<Bar>`
-	note: you might have meant to return the `match` expression
-	  --> $DIR/match-with-different-arm-types-as-stmt-instead-of-expr.rs:27:6
-	   |
-	LL |   fn wrong(c: &str) -> Box<dyn Foo> {
-	   |                        ------------ the `match` arms can conform to this return type
-	LL | /     match c {
-	LL | |         "baz" => Box::new(Baz),
-	LL | |         _ => Box::new(Bar),
-	LL | |     };
-	   | |     -^ the `match` is a statement because of this semicolon, consider removing it
-	   | |_____|
-	   |       this could be implicitly returned but it is a statement, not a tail expression
37	
38	error[E0308]: mismatched types
39	  --> $DIR/match-with-different-arm-types-as-stmt-instead-of-expr.rs:23:22

```

Reply:
maybe there is another place where to fix this error?

Reply:
ah nevermind, you changed the logic above this piece and I just didn't check the comment properly. This is resolved now. This specific `find_by_def_id` was correct, but at the time when this comment was written the `hir.maybe_body_owned_by(self.body_id)` was something else still